format/formatter.go

```diff
@@ -47,6 +47,132 @@ type Formatter interface {
 	Close() error
 }
 
+type anyRecord struct {
+	v interface{}
+}
+
+func (r *anyRecord) UnmarshalJSON(b []byte) error {
+	var i interface{}
+	err := json.Unmarshal(b, &i)
+	if err != nil {
+		return err
+	}
+
+	switch j := i.(type) {
+	case map[string]interface{}:
+		if d, ok := r.formatDate(j); ok {
+			r.v = d
+			return nil
+		}
+
+		r.traverseMap(&j)
+		r.v = j
+	default:
+		r.v = i
+	}
+
+	return err
+}
+
+func (r *anyRecord) traverseMap(i *map[string]interface{}) {
+	for k, v := range *i {
+		if m, ok := v.(map[string]interface{}); ok {
+			if d, ok := r.formatDate(m); ok {
+				(*i)[k] = d
+			} else {
+				r.traverseMap(&m)
+			}
+		}
+	}
+}
+
+// getAsInt returns the value as an int if it is a float64 or int
+func getAsInt(v interface{}) (int, bool) {
+	switch i := v.(type) {
+	case float64:
+		return int(i), true
+	case int:
+		return i, true
+	default:
+		return 0, false
+	}
+}
+
+// formatDate if i is recognized as a RethinkDb date, the date is returned as a RFC3339 formatted string
+func (r *anyRecord) formatDate(i map[string]interface{}) (string, bool) {
```

Comment:
Would it make sense to return an error here instead of a `bool`? In isolation, if you cannot format to a `string`, doesn't that indicate a failure?

Reply:
No, not all elements are dates. This function only converts dates and let others thru. The bool value just indicates if it was recognized as a date.

Reply:
I see, in that case I think I need to go through this part with you as well, as I am a bit confused about what this code does.
It seems like we are trying to match some datetime pattern, could a `regex` or two perhaps do the same task, but with a bit fewer lines?

Sometimes when I want to have specific error messages for what went wrong when parsing a string I will avoid `regex`, but it does not seem to be the case right here.

Reply:
Discussed this IRL: I did not read this function properly, everything is fine